Sources/Jobs/JobQueueHandler.swift

```diff
@@ -119,4 +130,70 @@ final class JobQueueHandler<Queue: JobQueueDriver>: Service {
 
 extension JobQueueHandler: CustomStringConvertible {
     public var description: String { "JobQueueHandler<\(String(describing: Queue.self))>" }
+    private var metricsLabel: String { "swift_jobs" }
+    private var meterLabel: String { "swift_jobs_meter" }
+
+    /// Used for the histogram which can be useful to see by job status
+    private enum JobStatus: String, Codable, Sendable {
+        case queued
+        case processing
+        case retried
+        case cancelled
+        case failed
+        case succeeded
+    }
+
+    private func updateJobMetrics(
+        for name: String,
+        startTime: UInt64,
+        error: Error? = nil,
+        retrying: Bool = false
+    ) {
+        
+        if retrying {
+            // dec processing when retried
+            Meter(label: self.meterLabel, dimensions: [("status", JobStatus.processing.rawValue)]).decrement()
+            Counter(
+                label: self.metricsLabel,
+                dimensions: [("name", name), ("status", JobStatus.retried.rawValue)]
+            ).increment()
+            return
+        }
+        
+        let jobStatus: JobStatus = if let error {
+            if error is CancellationError {
+                .cancelled
+            } else {
+                .failed
+            }
+        } else {
+            .succeeded
+        }
+        
+        let dimensions: [(String, String)] = [
+            ("name", name),
+            ("status", jobStatus.rawValue)
+        ]
+
+        // Calculate job execution time
+        Timer(
+            label: "\(self.metricsLabel)_duration_seconds",
+            dimensions: dimensions,
+            preferredDisplayUnit: .seconds
+        ).recordNanoseconds(DispatchTime.now().uptimeNanoseconds - startTime)
+
+        // Increment job counter base on status
+        Counter(
+            label: self.metricsLabel,
+            dimensions: dimensions
+        ).increment()
+        
+        self.updateJobMeters()
+    }
+    
+    private func updateJobMeters() {
```

Comment:
Can we rename this. It isn't clear what it is doing. Also if this is called if the job fails to decode, or it couldn't find the job id then processing is decremented when it hasn't been incremented.

I'm wondering if it be easier to break out the queued and processing counter editing from here and just place at the top of `runJob`. You'd also have to remove the processing `decrement` when you retry a job. 

```swift
Meter(label: self.meterLabel, dimensions: [("status", JobStatus.queued.rawValue)]).decrement()
Meter(label: self.meterLabel, dimensions: [("status", JobStatus.processing.rawValue)]).increment()
defer {
    Meter(label: self.meterLabel, dimensions: [("status", JobStatus.processing.rawValue)]).decrement()
}

Reply:
> Can we rename this. It isn't clear what it is doing. Also if this is called when the job fails to decode, or it couldn't find the job id then processing is decremented when it hasn't been incremented.
> 
> I'm wondering if it be easier to break out the queued and processing counter editing from here and just place at the top of `runJob`. You'd also have to remove the processing `decrement` when you retry a job.
> 
> ```swift
> Meter(label: self.meterLabel, dimensions: [("status", JobStatus.queued.rawValue)]).decrement()
> Meter(label: self.meterLabel, dimensions: [("status", JobStatus.processing.rawValue)]).increment()
> defer {
>     Meter(label: self.meterLabel, dimensions: [("status", JobStatus.processing.rawValue)]).decrement()
> }
> ```

This is much cleaner! I will update accordingly, thanks!

Reply:
@adam-fowler I updated the PR with the suggestion above.